src/stream/src/executor/sort.rs

```diff
@@ -0,0 +1,543 @@
+                        // Commit the epoch.
+                        self.state_table.commit(barrier.epoch).await?;
+                    } else {
+                        // If the barrier is not a checkpoint, then there is no actual data to
+                        // commit. Therefore, we simply update the epoch of state table.
+                        self.state_table.commit_no_data_expected(barrier.epoch);
+                    }
+
+                    // Update the vnode bitmap for the state table if asked. Also update the buffer.
+                    if let Some(vnode_bitmap) = barrier.as_update_vnode_bitmap(self.context.id) {
+                        let prev_vnode_bitmap =
+                            self.state_table.update_vnode_bitmap(vnode_bitmap.clone());
+                        self.fill_buffer(Some(&prev_vnode_bitmap), &vnode_bitmap)
+                            .await?;
+                    }
+
+                    yield Message::Barrier(barrier);
+                }
+            }
+        }
+    }
+
+    /// Fill the buffer when the sort executor initializes or scales. On initialization (including
+    /// recovering), `prev_vnode_bitmap` will be `None`. On scaling, `prev_vnode_bitmap` should be
+    /// set as the previous vnode bitmap to perform buffer update. Note that we do not assume
+    /// set relations between `prev_vnode_bitmap` and `curr_vnode_bitmap` when scaling occurs. That
+    /// is to say, `prev_vnode_bitmap` does not necessarily contain `curr_vnode_bitmap` on scaling
+    /// out, and vice versa. Therefore, we always check vnodes that are no longer owned and newly
+    /// owned by the sort executor regardless of the scaling type (scale in or scale out).
+    async fn fill_buffer(
+        &mut self,
+        prev_vnode_bitmap: Option<&Bitmap>,
+        curr_vnode_bitmap: &Bitmap,
+    ) -> StreamExecutorResult<()> {
+        // When scaling occurs, we remove data with vnodes that are no longer owned by this executor
+        // from buffer.
+        if let Some(prev_vnode_bitmap) = prev_vnode_bitmap {
+            let no_longer_owned_vnodes =
+                Bitmap::bit_saturate_subtract(prev_vnode_bitmap, curr_vnode_bitmap);
+            self.buffer.retain(|(_, pk), _| {
+                let vnode = self.state_table.compute_vnode(pk);
+                !no_longer_owned_vnodes.is_set(vnode as _)
+            });
+        }
+
+        // Read data with vnodes that are newly owned by this executor from state store. This is
+        // performed both on initialization and on scaling.
+        let newly_owned_vnodes = if let Some(prev_vnode_bitmap) = prev_vnode_bitmap {
+            Bitmap::bit_saturate_subtract(curr_vnode_bitmap, prev_vnode_bitmap)
+        } else {
+            curr_vnode_bitmap.to_owned()
+        };
+        let mut values_per_vnode = Vec::new();
+        for (owned_vnode, _) in newly_owned_vnodes
+            .iter()
+            .filter(|is_set| *is_set)
+            .enumerate()
```

Comment:
May add a new function on `Bitmap` to iterate on all set bits.

Reply:
I have added this function in the WatermarkFilter branch. I'll modify this to use the function in my PR.